### PR TITLE
Fix: sort document reference by long type id

### PR DESCRIPTION
**Pull request by milaGGL**
Document ID supports strings and well as long integers in the format of "id" . When sorting documents by document ID, it should be sorted in the following order:
1. Long (numeric order)
2. String (lexicographic order)

#no-changelog

### CODE REVIEW

File: Firestore/core/src/model/base_path.h [hunk 1]
```diff
@@ -150,8 +150,18 @@ class BasePath {
            std::equal(begin(), end(), potential_child.begin());
   }
 
+  /**
+   * Compares the current path against another Path object. Paths are compared
+   * segment by segment, prioritizing numeric IDs (e.g., "__id123__") in numeric
+   * ascending order, followed by string segments in lexicographical order.
+   */
   util::ComparisonResult CompareTo(const T& rhs) const {
-    return util::CompareContainer(segments_, rhs.segments_);
```
Comment (milaGGL) [inline Firestore/core/src/model/base_path.h:154]: CompareContainer is a generic method that compares items in 2 T& variables by iteration, and it is only used here in base_path. The sorting order is unique to base_path(specifically resource_path), makes more sense to add the logic here instead of modifying the CompareContainer function.

File: Firestore/core/src/model/base_path.h [hunk 2]
```diff
@@ -174,6 +184,31 @@ class BasePath {
 
  private:
   SegmentsT segments_;
+
+  static util::ComparisonResult compareSegments(const std::string& lhs,
```
Comment (ehsannas) [inline Firestore/core/src/model/base_path.h:188]: C++ [style](https://google.github.io/styleguide/cppguide.html#Function_Names) guide suggests function names to start with a capital letter (`CompareSegments`)

File: Firestore/core/src/model/base_path.h [hunk 3]
```diff
@@ -174,6 +184,31 @@ class BasePath {
...
+    } else if (isLhsNumeric && isRhsNumeric) {
+      return util::Compare(extractNumericId(lhs), extractNumericId(rhs));
+    } else {
+      return util::Compare(lhs, rhs);
+    }
+  }
+
+  static bool isNumericId(const std::string& segment) {
+    return segment.substr(0, 4) == "__id" &&
+           segment.substr(segment.size() - 2) == "__";
```
Comment (ehsannas) [inline Firestore/core/src/model/base_path.h:206]: `substr` throws and exception if `pos` is larger than the string length ([source](https://en.cppreference.com/w/cpp/string/basic_string/substr)). Also not sure what the behavior would be if segment.size is 1 (and therefore segment.size()-2 will be -1). Please add a unit test for these corner cases.

You may want to do:
```cpp
return
  segment.size() > 6 &&
  segment.substr(0, 4) == "__id" &&
  segment.substr(segment.size() - 2) == "__";
```

File: Firestore/core/src/model/base_path.h [hunk 4]
```diff
@@ -174,6 +184,31 @@ class BasePath {
...
+  }
+
+  static int64_t extractNumericId(const std::string& segment) {
+    return std::stol(segment.substr(4, segment.size() - 2));
```
Comment (ehsannas) [inline Firestore/core/src/model/base_path.h:210]: `std::stol` throws an `std::invalid_argument` exception if the argument is an empty string. So it reinforces that this function should *not* be called if the segment length is 6 or smaller.

**Review by ehsannas — CHANGES_REQUESTED** [review]

File: Firestore/core/src/model/base_path.h (hunk 3 above)
Comment (ehsannas) [reply Firestore/core/src/model/base_path.h:206]: by adding unit tests I mean testing that the output of `isNumericId` is `false` for (`"__id"` or `"__id__"` or `""`) for example.

Comment (milaGGL) [reply Firestore/core/src/model/base_path.h:206]: The test cases include documentIds like "12", "7", "A", "Aa", whose length is less than 6 and IsNumericId returns false.

Comment (milaGGL) [reply Firestore/core/src/model/base_path.h:206]: ""  and "__id__" are not allowed.

**Review by ehsannas — APPROVED** [review]
